From 3226f9caddff492a26d56d1d1ead7a2d6e9754b2 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Tue, 5 Nov 2024 20:09:45 +0100
Subject: [PATCH 1354/1695] plat-stm32mp1: stm32mp1_pwr: fix wake-up pins

The PWR wake-up pin are used as wake up interruption and also
as normal interruption, so the API stm32_exti_unmask is required
to set the interruption bit in EXTI IMR  register.

This code is aligned with the last stm32mp25_pwr behavior.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 959174c581658a161b30234173b631076ad4d007 plat-stm32mp1: stm32mp1_pwr: fix wake-up pins]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I21e726397d6d088e888a13425e89c036ef01d5b3
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/416819
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
index 0aecc4e9f..a84e4f97c 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
@@ -212,8 +212,10 @@ static void stm32mp1_pwr_itr_enable_nolock(size_t it)
 
 	VERBOSE_PWR("Pwr irq enable");
 
-	if (IS_ENABLED(CFG_STM32_EXTI))
+	if (IS_ENABLED(CFG_STM32_EXTI)) {
+		stm32_exti_unmask(priv->exti, PWR_EXTI_WKUP1 + it);
 		stm32_exti_enable_wake(priv->exti, PWR_EXTI_WKUP1 + it);
+	}
 
 	io_setbits32(priv->base + MPUWKUPENR, BIT(it));
 }
@@ -226,8 +228,10 @@ static void stm32mp1_pwr_itr_disable_nolock(size_t it)
 
 	io_clrbits32(priv->base + MPUWKUPENR, BIT(it));
 
-	if (IS_ENABLED(CFG_STM32_EXTI))
+	if (IS_ENABLED(CFG_STM32_EXTI)) {
+		stm32_exti_mask(priv->exti, PWR_EXTI_WKUP1 + it);
 		stm32_exti_disable_wake(priv->exti, PWR_EXTI_WKUP1 + it);
+	}
 }
 
 static TEE_Result stm32_pwr_irq_set_trig(size_t it, unsigned int flags)
-- 
2.47.2

